File: ofSketchApp/src/Project.cpp
// =============================================================================
//
// Copyright (c) 2013-2014 Christopher Baker <http://christopherbaker.net>
//               2014 Brannon Dorsey <http://brannondorsey.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//
// =============================================================================


#include "Project.h"
#include "ofUtils.h"


namespace of {
namespace Sketch {


const std::string Project::SKETCH_FILE_EXTENSION = "sketch";


Project::~Project()
{
}


Project::Project(const std::string& path): _path(path), _isLoaded(false)
{
    // this is not efficient at all! I am just keeping these FileTemplate loads in the project
    // constructor because it makes the most sense architecure wise.
    _classFileTemplate = ofBufferFromFile(ofToDataPath("Resources/Templates/SketchTemplates/class.tmpl")).getText();
    load(_path, getName());
}

void Project::load(const std::string& path, const std::string& name)
{
    _sketchDir = ofDirectory(ofToDataPath(path + "/sketch"));
    
    _data.clear();
    
    if (_sketchDir.exists()) 
    {
        _sketchDir.listDir();

        std::vector<ofFile> files = _sketchDir.getFiles();

        int classCounter = 0;

        for (std::size_t i = 0; i < files.size(); ++i) 
        {
            ofFile file = files[i];

            if (file.getBaseName() == name)
            {
                file.open(file.getAbsolutePath());
                _data["projectFile"]["name"] = file.getBaseName();
                _data["projectFile"]["fileName"] = file.getFileName();
                _data["projectFile"]["fileContents"] = file.readToBuffer().getText();
            } 
            else if (file.getExtension() == SKETCH_FILE_EXTENSION)
            {
                file.open(file.getAbsolutePath());
                _data["classes"][classCounter]["name"] = file.getBaseName();
                _data["classes"][classCounter]["fileName"] = file.getFileName();
                _data["classes"][classCounter]["fileContents"] = file.readToBuffer().getText();
                classCounter++;
            }
        }
        
        _loadAddons();
        _isLoaded = true;
    }
}


bool Project::isLoaded() const
{
    return _isLoaded;
}

// saves differences only
void Project::save(const Json::Value& data)
{
    // this is not working for some reason...
    if (_data != data)
    {
        ofLogVerbose("Project::save") << data.toStyledString();
        
        if (_data["projectFile"] != data["projectFile"])
        {
            _data["projectFile"] = data["projectFile"];
            _saveFile(_data["projectFile"]);
        }
        
            
        // uses nested for loop in case classes are not in the same order
        std::vector<Json::Value> newClasses;
        std::vector<Json::Value> deletedClasses;
        

        for (unsigned int i = 0; i < _data["classes"].size(); ++i)
        {
            Json::Value& classFile = _data["classes"][i];

            bool matchFound = false;
            
            for (unsigned int j = 0; j < data.size(); ++j)
            {
                const Json::Value& newClassFile = data["classes"][j];
                
                if (classFile["fileName"] == newClassFile["fileName"]) 
                {
                    if (classFile != newClassFile)
                    {
                        classFile = newClassFile;
                        _saveFile(classFile);
                    }

                    matchFound = true;
                    break;
                }
            }
            
            if (!matchFound) _saveFile(classFile); // class is new
        }
    }
    else
    {
        ofLogNotice("Project::save") << "Project data is the same. Not saving project.";
    }
}


bool Project::create(const std::string& path)
{
    ofDirectory project(ofToDataPath(path));

    if (!project.exists()) 
    {
        ofDirectory temp(ofToDataPath("Resources/Templates/SimpleTemplate"));
        temp.copyTo(ofToDataPath(path));
        return true;
    }
    
    return false;
}


bool Project::remove()
{
    ofDirectory projectDir(getPath());
    return projectDir.remove(true);
}


bool Project::rename(const std::string& newName)
{
    if (isLoaded()) 
    {
        ofLogVerbose("Project::rename") << "renaming project \"" << getName() << "\" to \"" << newName + "\"";

        ofFile projectDir(getPath());

        std::string oldProjectName = getName();

        ofLogVerbose("Project::rename") << "project path: " << getPath() << " Renamed to: " << projectDir.getEnclosingDirectory() + newName;

        if (!projectDir.renameTo(projectDir.getEnclosingDirectory() + newName)) return false;

        _path = projectDir.getAbsolutePath();

        ofFile projectFile(projectDir.getAbsolutePath() + "/sketch/" + oldProjectName + "." + SKETCH_FILE_EXTENSION);
        
        
        // remove old executable
        ofDirectory bin;
        bin.listDir(projectDir.path() + "/bin");
        std::vector<ofFile> files = bin.getFiles();
        
        Poco::RegularExpression appExpression( oldProjectName + "(.exe|.app)*$", Poco::RegularExpression::RE_ANCHORED);

        for (int i = 0; i < files.size(); i++)
        {
            std::string baseName = files[i].getBaseName();
            if (appExpression.match(baseName)) {
                files[i].remove(true);
                ofLogVerbose("Project::rename")
                    << "removed " << files[i].getFileName() << " from " << bin.getAbsolutePath() << endl;
            }
        }
        

        ofLogVerbose("Project::rename") << "projectDir path after rename: " << projectDir.getAbsolutePath();
        ofLogVerbose("Project::rename") << "projectFile path: " << projectFile.getAbsolutePath();

        if (!projectFile.renameTo(projectDir.getAbsolutePath() + "/sketch/" + newName + "." + SKETCH_FILE_EXTENSION)) return false;

        ofLogVerbose("Project::rename") << "projectFile path after rename: " << projectFile.getAbsolutePath();

        _data["projectFile"]["name"] = newName;

        _data["projectFile"]["fileName"] = newName + "." + SKETCH_FILE_EXTENSION;

        return true;
    }
    else
    {
        ofLogVerbose("Project::rename") << "Cannot rename project, it is not loaded.";
        return false;
    }
    
}


Json::Value Project::createClass(const std::string& className)
{
    std::string fileContents = _classFileTemplate;
    ofStringReplace(fileContents, "<classname>", className);

    ofLogVerbose("Project::createClass") << "fileContents: "<< fileContents;

    Json::Value classFile;
    // TODO: Load extension from settings
    classFile["fileName"] = className + "." + SKETCH_FILE_EXTENSION;
    classFile["name"] = className;
    classFile["fileContents"] = fileContents;
    _data["classes"][getNumClasses()] = classFile;
    // TODO: re-loading is a terribly slow way to delete. Come back and optimize.
    // Simply need to remove the Json::Value class in _data["classes"]
    _saveFile(classFile);
    return classFile;
}


bool Project::deleteClass(const std::string& className)
{
    if (isLoaded())
    {
        ofFile file(_sketchDir.getAbsolutePath() + "/" + className + "." + SKETCH_FILE_EXTENSION);

        if (file.exists())
        {
            file.remove();
            // TODO: re-loading is a terribly slow way to delete. Come back and optimize.
            // Simply need to remove the Json::Value class in _data["classes"]
            load(_path, getName());
            return true;
        }
    }

    return false;
}

bool Project::renameClass(const std::string& currentName, const std::string& newName)
{
    if (isLoaded())
    {
        ofLogVerbose("Project::renameClass") << "Renaming class...";

        ofFile file(_sketchDir.getAbsolutePath() + "/" + currentName + "." + SKETCH_FILE_EXTENSION);

        if (file.exists() && hasClasses()) 
        {
            unsigned int numClasses = getNumClasses();

            for (unsigned int i = 0; i < numClasses; ++i)
            {
                if (_data["classes"][i]["name"] == currentName)
                {
                    _data["classes"][i]["name"] = newName;
                    _data["classes"][i]["fileName"] = newName + "." + SKETCH_FILE_EXTENSION;

                    file.renameTo(_sketchDir.getAbsolutePath() + "/" + newName + "." + SKETCH_FILE_EXTENSION);
                    return true;
                }
            }
        }
    }

    return false;
}


bool Project::hasClasses() const
{
    return !_data["classes"].empty();
}


bool Project::isClassName(const std::string& className) const
{
    if (hasClasses()) 
    {
        unsigned int numClasses = getNumClasses();

        for (unsigned int i = 0; i < numClasses; ++i)
        {
            if (_data["classes"][i]["name"] == className)
            {
                return true;
            }
        }
    }
    
    return false;
}

    
unsigned int Project::getNumClasses() const
{
    return _data["classes"].size();
}


const std::string& Project::getPath() const
{
    return _path;
}


std::string Project::getName() const
{
    ofFile project(_path);
    return project.getBaseName();
}


const Json::Value& Project::getData() const
{
    return _data;
}
    
void Project::addAddon(std::string& addon)
{
    if (!usingAddon(addon)) {
        _addons.push_back(addon);
        _saveAddons();
    }
}
    
bool Project::removeAddon(std::string& addon)
{
    for (unsigned int i = 0; i < _addons.size(); i++)
    {
        if (addon == _addons[i]) {
            _addons.erase(_addons.begin() + i);
            _saveAddons();
            return true;
        }
    }
    
    return false;
}
  
bool Project::hasAddons() const
{
    return _addons.size() > 0;
}
    
bool Project::usingAddon(std::string& addon) const
{
    for (unsigned int i = 0; i < _addons.size(); i++)
    {
        if (addon == _addons[i]) {
            return true;
        }
    }
    
    return false;
}
    
std::vector<std::string> Project::getAddons() const
{
    return _addons;
}
    
void Project::_loadAddons()
{
    _addons.clear();
    ofFile addonsMakefile(_path + "/addons.make");
    
    if (addonsMakefile.exists()) {
        
        std::vector<std::string> lines = ofSplitString(ofBufferFromFile(addonsMakefile.path()), "\n");
        
        for (unsigned int i = 0; i < lines.size(); i++) {
            
            if (lines[i] != "") {
                
                _addons.push_back(lines[i]);
            }
        }
    }
}
    
void Project::_saveAddons()
{
    ofBuffer buffer(ofJoinString(_addons, "\n"));
    ofBufferToFile(_path + "/addons.make", buffer);
}

void Project::_saveFile(const Json::Value& fileData)
{
    ofBuffer fileBuffer(fileData["fileContents"].asString());
    ofBufferToFile(getPath() + "/sketch/" + fileData["fileName"].asString(), fileBuffer);
}

} } // namespace of::Sketch

File: ofSketchApp/bin/data/DocumentRoot/js/ofSketch/classes/Project.js
// =============================================================================
//
// Copyright (c) 2014 Brannon Dorsey <http://brannondorsey.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//
// =============================================================================

/*
_data = {
    "projectFile": {
        "name": "myProject"
        "fileName": "myProject.sketch",
        "fileContents": ""
    },
    "classes": [
        {
            "name": "myProject"
            "fileName": "myClass.sketch",
            "fileContents": ""
        }
    ]
}
 */

function Project(projectName, onSuccess, onError, isTemplate) {
  var _self = this;
  var _needsSave = false;
  var _data;
  var _isTemplate;

  this.load = function(projectName, onSuccess, onError) {
    JSONRPCClient.call('load-project', {
        projectName: projectName
      },
      function(result) {
        _data = result.data;
        onSuccess(result);
      },
      function(error) {
        onError(error);
      });
  }

  this.loadTemplate = function(onSuccess, onError) {
    JSONRPCClient.call('load-template-project', {},
      function(result) {
        _data = result;
        onSuccess(result);
      },
      function(error) {
        onError(error);
      });
  }

  // used only if isAnonymous()
  this.create = function(onSuccess, onError) {
    JSONRPCClient.call('create-project', {
        projectData: _self.getData()
      },
      function(result) {
        _isTemplate = false;
        onSuccess(result);
      },
      function(error) {
        onError(error);
      });
  }

  this.save = function(onSuccess, onError) {
    JSONRPCClient.call('save-project', {
        projectData: _self.getData()
      },
      function(result) {
        _needsSave = false;
        onSuccess(result);
      },
      function(error) {
        onError(error);
      });
  }

  this.rename = function(newProjectName, onSuccess, onError) {
    JSONRPCClient.call('rename-project', {
        projectName: _self.getName(),
        newProjectName: newProjectName,
        clientUUID: CLIENT_UUID
      },
      function(result) {
        _data.projectFile.name = newProjectName;
        _data.projectFile.fileName = newProjectName + ".sketch";
        onSuccess(result);
      },
      function(err) {
        console.log("Rename project error:");
        console.log(err);
        onError(err);
      });
  }

  this.createClass = function(className, onSuccess, onError) {
    JSONRPCClient.call('create-class', {
        projectName: _self.getName(),
        className: className
      },
      function(result) {
        var classFile = result.classFile;
        if (!_self.hasClasses()) _data.classes = [];
        _self.getClasses().push(classFile);
        onSuccess(classFile);
      },
      function(error) {
        onError(error);
      });
  }

  this.deleteClass = function(className, onSuccess, onError) {
    JSONRPCClient.call('delete-class', {
        projectName: _self.getName(),
        className: className
      },
      function(result) {
        _data.classes = _.without(_self.getClasses(), _.findWhere(_self.getClasses(), {
          name: className
        }));
        onSuccess(result);
      },
      function(error) {
        onError(error);
      });
  }

  this.renameClass = function(className, newClassName, onSuccess, onError) {
    JSONRPCClient.call('rename-class', {
        projectName: _self.getName(),
        className: className,
        newClassName: newClassName
      },
      function(result) {

        var c = _.findWhere(_self.getClasses(), {
          name: className
        });
        c.name = newClassName;
        c.fileName = newClassName + '.sketch';
        onSuccess(result);
      },
      function(error) {
        onError(error);
      });
  }

  this.getAddonList = function(onSuccess, onError) {
    JSONRPCClient.call('get-project-addon-list', {
        projectName: _self.getName(),
      },
      function(result) {

        var addons = [];

        if (result.hasAddons) {
          addons = result.addons;
        }

        onSuccess(addons);
      },
      function(error) {
        onError(error);
      });
  }

  this.addAddon = function(addon, onSuccess, onError) {
    JSONRPCClient.call('add-project-addon', {
        projectName: _self.getName(),
        addon: addon
      },
      function(result) {
        onSuccess(result);
      },
      function(error) {
        onError(error);
      });
  }

  this.removeAddon = function(addon, onSuccess, onError) {
    JSONRPCClient.call('remove-project-addon', {
        projectName: _self.getName(),
        addon: addon
      },
      function(result) {
        onSuccess(result);
      },
      function(error) {
        onError(error);
      });
  }

  // used to assign the name of a template project only
  this.assignName = function(name) {
    if (_self.isTemplate()) {
      _data.projectFile.name = name;
      _data.projectFile.fileName = name + ".sketch";
    }
  }

  this.getName = function() {
    return _data.projectFile.name;
  }

  this.isClassName = function(className) {
    return _self.hasClasses() &&
      !_.isUndefined(_.findWhere(_self.getClasses(), {
        name: className
      }));
  }

  this.isTemplate = function() {
    return _isTemplate;
  }

  this.getProjectFile = function() {
    return _data.projectFile;
  }

  this.hasClasses = function() {
    return !_.isUndefined(_data.classes) && !_.isNull(_data.classes);
  }

  this.getClasses = function() {
    return _data.classes;
  }

  this.addClassFile = function(c) {
    _data.classes.push(c);
  }

  this.getData = function() {
    return _data;
  }

  this.needsSave = function() {
    return _needsSave;
  }

  this.setNeedsSave = function(bool) {
    _needsSave = bool
  }

  if (isTemplate) {
    _self.loadTemplate(onSuccess, onError);
    _isTemplate = true;
  } else {
    _self.load(projectName, onSuccess, onError);
    _isTemplate = false;
  }
}
